### PR TITLE
ARROW-18162: [C++] Add Arm SVE compiler options

### PR DESCRIPTION
**Pull request by cyb70289**
As xsimd only supports fixed-size SVE, we have to specify vector size
explicitly on command line. And the binary can only run on hardware
with matched vector size. Otherwise, the code behaviour is undefined.
E.g., `cmake -DARROW_SIMD_LEVEL=SVE256 ..`
According macro `ARROW_HAVE_SVE256` and cmake variable are defined.

We can also leverage compiler auto vectorization to generate size
agnostic SVE code without specifying the vector size.
E.g., `cmake -DARROW_SIMD_LEVEL=SVE ..`

This PR also removes some unused Arm64 arch options.

### CODE REVIEW

File: cpp/cmake_modules/SetupCxxFlags.cmake [hunk 1]
```diff
@@ -496,14 +494,24 @@ if(ARROW_CPU_FLAG STREQUAL "armv8")
     endif()
     set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} ${ARROW_ARMV8_ARCH_FLAG}")
 
+    set(ARROW_HAVE_NEON ON)
```
Comment (pitrou) [inline cpp/cmake_modules/SetupCxxFlags.cmake:497]: Is NEON always available with SVE?

File: cpp/cmake_modules/SetupCxxFlags.cmake [hunk 2]
```diff
@@ -496,14 +494,24 @@ if(ARROW_CPU_FLAG STREQUAL "armv8")
...
-    if(ARROW_ARMV8_ARCH_FLAG MATCHES "\\+crypto")
-      add_definitions(-DARROW_HAVE_ARMV8_CRYPTO)
-    endif()
-    # armv8.1+ implies crc support
-    if(ARROW_ARMV8_ARCH_FLAG MATCHES "armv8\\.[1-9]|\\+crc")
-      add_definitions(-DARROW_HAVE_ARMV8_CRC)
+    if(ARROW_SIMD_LEVEL MATCHES "SVE[0-9]*")
+      if(NOT ARROW_ARMV8_ARCH_FLAG MATCHES "\\+sve")
+        message(FATAL_ERROR "SIMD Level is set to ${ARROW_SIMD_LEVEL} but SVE is not enabled in compiler options. "
+                            "Add \"-DARROW_ARMV8_ARCH=armv8-a+sve\" to cmake command line to enable SVE."
```
Comment (pitrou) [inline cpp/cmake_modules/SetupCxxFlags.cmake:503]: Is it useful to error on this? Why not automatically enable SVE when it's given in ARROW_SIMD_LEVEL?

File: python/CMakeLists.txt [hunk 3]
```diff
@@ -129,7 +129,7 @@ endif()
 if(NOT DEFINED ARROW_ARMV8_ARCH)
   set(ARROW_ARMV8_ARCH
       "armv8-a"
-      CACHE STRING "Arm64 arch and extensions: armv8-a, armv8-a or armv8-a+crc+crypto")
+      CACHE STRING "Arm64 arch and extensions")
```
Comment (pitrou) [inline python/CMakeLists.txt:132]: Why do we have this option separately in Python, by the way?

**Comment by pitrou** [comment]
It seems we could also use SVE in `src/arrow/util/bpacking.cc`, `src/arrow/util/utf8_internal.h`, as well as `src/arrow/csv/lexing_internal.h` and `src/arrow/csv/writer.cc`?

File: cpp/cmake_modules/SetupCxxFlags.cmake (hunk 1 above)
Comment (cyb70289) [reply cpp/cmake_modules/SetupCxxFlags.cmake:497]: Yes, NEON is always available for aarch64.

File: cpp/cmake_modules/SetupCxxFlags.cmake (hunk 2 above)
Comment (cyb70289) [reply cpp/cmake_modules/SetupCxxFlags.cmake:503]: I'm thinking about removing this cmake option `ARROW_ARMV8_ARCH`, and use only `ARROW_SIMD_LEVEL`.
https://github.com/apache/arrow/blob/master/cpp/cmake_modules/DefineOptions.cmake#L204

`ARROW_ARMV8_ARCH` is append to `gcc -march=???`, it's only for Arm. x86 doesn't need `-march` as x86 has compiler options for each feature, e.g., `-msse4.2`, `-mavx2`, etc, to enable specific SIMD instructions.
But Arm doesn't have these compiler options, CPU features are set by `-march` only, e.g., `-march=armv8-a+sve`, `-march=armv8.2-a+crypto`, etc.
Specify `-march` by cmake command line gives more flexibility, but looks most of the time it also introduces unnecessary complexity.

**Comment by cyb70289** [comment]
> It seems we could also use SVE in `src/arrow/util/bpacking.cc`, `src/arrow/util/utf8_internal.h`, as well as `src/arrow/csv/lexing_internal.h` and `src/arrow/csv/writer.cc`?

Yes. SVE may be better when longer vector size is beneficial, or to leverage operations (e.g, scatter/gather) not availble in NEON.

File: dev/tasks/conda-recipes/arrow-cpp/build-pyarrow.sh [hunk 4]
```diff
@@ -39,11 +39,6 @@ else
     export PYARROW_WITH_CUDA=0
 fi
 
-# Resolve: Make Error at cmake_modules/SetupCxxFlags.cmake:338 (message): Unsupported arch flag: -march=.
```
Comment (cyb70289) [inline dev/tasks/conda-recipes/arrow-cpp/build-pyarrow.sh:42]: @kou which tests shall we run against this change?

Comment (kou) [reply dev/tasks/conda-recipes/arrow-cpp/build-pyarrow.sh:42]: `conda-linux-*` and `conda-osx-*` but we should wait for #14102.

**Comment by cyb70289** [comment]
@github-actions crossbow submit conda-{linux,osx}-*

**Comment by cyb70289** [comment]
@github-actions crossbow submit conda-linux-*

**Comment by cyb70289** [comment]
@github-actions crossbow submit conda-osx-*

**Comment by cyb70289** [comment]
conda-osx-clang-py310/py37 build error looks not related.
https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=38774&view=logs&j=cf796865-97b7-5cd1-be8e-6e00ce4fd8cf&t=88ee2fb8-46fd-5c68-fde3-1c8d31ba2a5f&l=3570

**Review by kou — APPROVED** [review]
+1 for changes

I re-ran failed jobs.

**Review by pitrou — APPROVED** [review]

**Pull request merged**